### PR TITLE
Move atlantis-data volume to a separate PVC

### PR DESCRIPTION
**Pull request by BrunoTarijon**
## what

Remove the atlantis-data volume creation from the sts and add a separate PVC.

## why

We can not change the storage size without deleting the deployed sts.

## tests

I have tested the logic using the helm template function to generate the manifests with the volumeClaim interface and the deprecated .Values.dataStorage.

### CODE REVIEW

**Comment by GenPage** [comment]
@BrunoTarijon LGTM, can you bump the chart version? It can be just a patch bump.

**Comment by BrunoTarijon** [comment]
> @BrunoTarijon LGTM, can you bump the chart version? It can be just a patch bump.

@GenPage ok, done

**Comment by bsvartz** [comment]
@BrunoTarijon, @GMartinez-Sisti  - is it possible to add the option to edit on this file:

https://github.com/runatlantis/helm-charts/blob/f56c4caa0fe5c83052d4f0b42f67c0d7f0e9d122/charts/atlantis/templates/pvc.yaml 

spec:
  accessModes:
    - ReadWriteOnce 
to ReadWriteMany - in order to support the following feature? https://github.com/runatlantis/atlantis/issues/3795?

and this:
  volumeClaimTemplates:
  - metadata:
      name: atlantis-data
    spec:
      accessModes: ["ReadWriteOnce"] # Volume should not be shared by multiple nodes.

on this file:
https://github.com/runatlantis/helm-charts/blob/f56c4caa0fe5c83052d4f0b42f67c0d7f0e9d122/charts/atlantis/templates/statefulset.yaml 

Generally - i want to be able to share the same volume between multiple atlantis instances.

**Comment by BrunoTarijon** [comment]
@bsvartz I have added the requested feature.

@GenPage. Does it look good to you? Can we merge it?

**Comment by GMartinez-Sisti** [comment]
> @bsvartz I have added the requested feature.
> 
> @GenPage. Does it look good to you? Can we merge it?

I think the idea was to make it configurable in the values file. Leaving the default as it was before, but you configure what you want.

**Comment by jamengual** [comment]
@BrunoTarijon could you fix the conflicts? Thanks.

**Comment by matilote** [comment]
Any chance to resolve the conflicts and get this merged?

**Comment by BrunoTarijon** [comment]
Sorry for not seeing this for a while. I have made the custom access mode in the PVC and chose not to change the STS configuration for the access mode because it is marked as deprecated.

@jamengual I have resolved the conflicts.

**Comment by BrunoTarijon** [comment]
Hey guys (@jamengual @GenPage) just receive the Stale label, could you review this PR?

**Comment by jamengual** [comment]
@GMartinez-Sisti

**Comment by jamengual** [comment]
@BrunoTarijon we have unit test now, could you take a look?

**Comment by BrunoTarijon** [comment]
Done @jamengual

File: charts/atlantis/templates/pvc.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{{- $usingVolumeClaimInterface := not .Values.dataStorage }}
+{{- if and .Values.volumeClaim.enabled $usingVolumeClaimInterface }}
```
Comment (GMartinez-Sisti) [inline charts/atlantis/templates/pvc.yaml:2]: nit: same result
```suggestion
{{- $usingVolumeClaimInterface := not .Values.dataStorage }}
{{- if and .Values.volumeClaim.enabled ( not .Values.dataStorage ) }}
```

**Comment by GMartinez-Sisti** [comment]
@BrunoTarijon looks like something [needs to be fixed](https://github.com/runatlantis/helm-charts/actions/runs/8157809563/job/22298331658?pr=304). The error is:

> 4m33s       Warning   FailedCreate           statefulset/atlantis-wbp015wolz                              statefulset-controller        create Pod atlantis-wbp015wolz-0 in StatefulSet atlantis-wbp015wolz failed error: Pod "atlantis-wbp015wolz-0" is invalid: spec.containers[0].volumeMounts[0].name: Not found: "atlantis-data"   10m          17      atlantis-wbp015wolz.17b9e455ea79e80f

Please adjust the ci files on https://github.com/runatlantis/helm-charts/tree/main/charts/atlantis/ci and create a new one to include your change. The only requirement is that filenames end with `*-values.yaml`.

**Comment by BrunoTarijon** [comment]
> Please adjust the ci files on https://github.com/runatlantis/helm-charts/tree/main/charts/atlantis/ci and create a new one to include your change. The only requirement is that filenames end with *-values.yaml.

Don't get it. This feat should work with the default values.

**Comment by GMartinez-Sisti** [comment]
> > Please adjust the ci files on https://github.com/runatlantis/helm-charts/tree/main/charts/atlantis/ci and create a new one to include your change. The only requirement is that filenames end with *-values.yaml.
> 
> Don't get it. This feat should work with the default values.
> 
> cc: @GMartinez-Sisti

Fair enough. Now the tests are passing.

**Review by GMartinez-Sisti — APPROVED** [review]

**Comment by cilindrox** [comment]
This should've been a major release - breaks existing installs using the STS pv

**Comment by bergemalm** [comment]
Yep, breaking change!

**Comment by GMartinez-Sisti** [comment]
> This should've been a major release - breaks existing installs using the STS pv

> Yep, breaking change!

In hindsight, you are both right. Our tests install everything from scratch, but they don't contemplate upgrades.

I tested by installing `4.21.1` and then tried to upgrade to `4.22.0` and got an error:

```sh
→ helm install atlantis runatlantis/atlantis --version 4.21.1
NAME: atlantis
LAST DEPLOYED: Fri Apr  5 18:59:30 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
2. Atlantis will not start successfully unless at least one of the following sets of credentials are specified (see values.yaml for detailed usage):
  - github
  - githubApp
  - gitlab
  - bitbucket
```

```sh
→ helm upgrade atlantis runatlantis/atlantis --version 4.22.0
W0405 18:59:46.772125   48368 warnings.go:70] unknown field "labels"
Error: UPGRADE FAILED: cannot patch "atlantis" with kind StatefulSet: StatefulSet.apps "atlantis" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden
```

We might create a new major release with update steps to fix this. Discussing with the other maintainers. 🙏

**Pull request merged**